gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -590,6 +590,15 @@ ApplicationId setupAndSubmitApplication() throws IOException, YarnException, Int
     amContainerLaunchContext.setEnvironment(YarnHelixUtils.getEnvironmentVariables(this.yarnConfiguration));
     amContainerLaunchContext.setCommands(Lists.newArrayList(buildApplicationMasterCommand(applicationId.toString(), resource.getMemory())));
 
+    if (this.jarCacheEnabled) {
+      Path jarCachePath = YarnHelixUtils.calculateJarCachePath(this.config);
+      // Retain at least the current and last month's jars to handle executions running for ~30 days max
+      boolean cleanedSuccessfully = YarnHelixUtils.retainKLatestJarCachePaths(jarCachePath.getParent(), 2, this.fs);
```

Comment:
does this run before or after caching jars?  e.g. do we save only two prior AND THEN potentially add one more or we've added any new one already prior to retention paring it down to two?

Reply:
It runs after caching the jars. But it uses a consistent `YARN_APPLICATION_LAUNCHER_START_TIME_KEY` in the job so no matter how many times we look at the cache path it's only creating one path at most, and that path would be the ones where the jars are being uploaded.

Reply:
would be great to add this info to a code comment here (in the caller, since it's the combination of the two being used, so less effective in either's javadoc).  the crux is that retention of K=2 won't save us in cases where K=3 might exceed any FS quotas.